### PR TITLE
feat(Template): allow template functions to return a React element

### PR DESCRIPTION
**Pull request by VonD**
about [#45](https://github.com/algolia/instantsearch.js/issues/45)

### CODE REVIEW

File: src/components/Template.js [hunk 1]
```diff
@@ -27,13 +27,16 @@ function Template(props) {
...
   return (
     <div
       {...otherProps}
       className={props.cssClass}
-      dangerouslySetInnerHTML={{__html: content}}
-    />
+      dangerouslySetInnerHTML={isElement ? null : {__html: content}}
+    >
+      {isElement ? content : null}
+    </div>
```
Comment (bobylito) [inline src/components/Template.js]: Maybe we should not systematically use dangerouslySetInnerHTML. What about:

```js
if(isEelement) {
  return <div...>{content}</div>
}
return <div...  dangerouslySetInnerHTML={{__html: content}}/>
```

Comment (VonD) [inline src/components/Template.js]: @bobylito ok, makes sense, i've pushed the change.

**Comment by vvo** [comment]
Somehow the browsers cannot start, this is a test build issue. The PR is nice

**Comment by vvo** [comment]
Thanks @VonD, a lot. Awesome contribution, where are you using/planning to use instantsearch.js? Any feedback on how in integrates with an existing react applications is very nice to have, open issues!

**Comment by VonD** [comment]
@vvo thanks! Not much feedback for the moment as the project is still at a very early stage and as i'm just discovering the lib.

I was wondering, though, if there might be a more straightforward integration within a React app. Excuse-me if it's already the case. I was thinking about an api like, for example : 

```es6
import React, { Component } from 'react' 
import { InstantSearch, SearchBox, Hits } from 'instantsearch.js'

class MySearchComponent extends Component {
  render() {
    return (
      <InstantSearch
        appId={ appId }
        apiKey={ apiKey }
        indexName={ indexName }
      >
        <SearchBox
          placeholder={ 'Custom placeholder' }
        />
        <Hits
          hitsPerPage={ 6 }
          renderHit={ hit => <MyItemComponent item={ hit } /> }
        />
      </InstantSearch>
    );
  }
}
```

Some of the benefits i see in such an approach would be : 
* the whole tree gets re-rendered when the parent component's `props` or `state` are updated
* the `context` does not get lost : useful with libs relying on it such as `react-router`
* we get server-side rendering for free
* no risk of conflicts between different React versions
* maybe modularize dependencies to avoid bundling `hogan` or other libs that react apps won't use

Would that be a big refactoring ? I understand the lib is not mainly targeting React apps, but as it is all React, it doesn't seem too difficult. The reference to the search helper could be passed to the widgets via `context`

Sorry if all that's already possible or off-topic.

**Comment by vvo** [comment]
@VonD can you open an new issue about this subject (just copy paste it)?

**Comment by VonD** [comment]
@vvo done

**Pull request merged**